File: src/main/java/co/uk/isxander/evergreenhud/elements/impl/ElementMemory.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/Evergreen-Client/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package co.uk.isxander.evergreenhud.elements.impl;

import co.uk.isxander.evergreenhud.elements.Element;
import co.uk.isxander.evergreenhud.settings.impl.ArraySetting;
import co.uk.isxander.evergreenhud.settings.impl.BooleanSetting;
import co.uk.isxander.xanderlib.utils.MathUtils;
import co.uk.isxander.evergreenhud.elements.ElementData;
import net.minecraftforge.fml.common.gameevent.TickEvent;

import java.text.DecimalFormat;

public class ElementMemory extends Element {

    private String memDisplay = "";
    private long lastUpdated = 0L;

    public BooleanSetting trailingZeros;
    public ArraySetting displayMode;

    @Override
    public void initialise() {
        addSettings(displayMode = new ArraySetting("Display", "How the value will be displayed.", "Absolute", new String[]{"Absolute", "Percentage"}));
        addSettings(trailingZeros = new BooleanSetting("Trailing Zeros", "Add zeroes to match the accuracy.", false));
    }

    @Override
    public ElementData metadata() {
        return new ElementData("Memory", "Shows how much memory minecraft is utilising.");
    }

    @Override
    protected String getValue() {
        return memDisplay;
    }

    // Update memory every second to minimize lag
    @Override
    public void onClientTick(TickEvent.ClientTickEvent event) {
        if (lastUpdated < System.currentTimeMillis() - 1000L) {
            if (displayMode.get().equalsIgnoreCase("absolute")) {
                DecimalFormat df = new DecimalFormat((trailingZeros.get() ? "0.0" : "#.#"));
                memDisplay = df.format(bytesToMb(Runtime.getRuntime().totalMemory() -
                        Runtime.getRuntime().freeMemory()) / 1024f) + " GB";
            } else {
                DecimalFormat df = new DecimalFormat((trailingZeros.get() ? "0.0%" : "#.#") + (displayMode.get().equalsIgnoreCase("percentage") ? "%" : " GB"));
                memDisplay = df.format(MathUtils.getPercent(bytesToMb(Runtime.getRuntime().totalMemory() -
                        Runtime.getRuntime().freeMemory()), 0, bytesToMb(Runtime.getRuntime().maxMemory())));
            }
            lastUpdated = System.currentTimeMillis();
        }
    }

    private long bytesToMb(long bytes) {
        return bytes / 1024L / 1024L;
    }

    @Override
    public String getDisplayTitle() {
        return "Mem";
    }

}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/impl/ElementMemory.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.impl

import dev.isxander.evergreenhud.elements.type.SimpleTextElement
import dev.isxander.evergreenhud.utils.decimalFormat
import dev.isxander.evergreenhud.utils.elementmeta.ElementMeta
import dev.isxander.evergreenhud.utils.getPercent
import dev.isxander.settxi.impl.OptionContainer
import dev.isxander.settxi.impl.boolean
import dev.isxander.settxi.impl.option

@ElementMeta(id = "evergreenhud:memory", name = "Memory", description = "Show the amount of RAM being used by Minecraft.", category = "Simple")
class ElementMemory : SimpleTextElement("Memory", 20) {
    var displayType by option(DisplayType.Absolute) {
        name = "Display Type"
        description = "How to display the amount of RAM"
        category = "Memory"
    }

    var trailingZeros by boolean(false) {
        name = "Trailing Zeros"
        description = "Make the number the same width."
        category = "Memory"
    }

    override fun calculateValue(): String {
        return when (displayType) {
            DisplayType.Absolute -> {
                val df = decimalFormat(1, trailingZeros)

                df.format(bytesToMb(Runtime.getRuntime().totalMemory() -
                        Runtime.getRuntime().freeMemory()) / 1024f) + " GB"
            }
            DisplayType.Percentage -> {
                val df = decimalFormat(1, trailingZeros, true)

                df.format(getPercent(bytesToMb(Runtime.getRuntime().totalMemory() -
                        Runtime.getRuntime().freeMemory()), 0, bytesToMb(Runtime.getRuntime().maxMemory())))
            }
            else -> "0"
        }
    }

    private fun bytesToMb(bytes: Long): Long {
        return bytes / 1024L / 1024L
    }

    object DisplayType : OptionContainer() {
        val Absolute = option("Absolute")
        val Percentage = option("Percentage")
    }
}
